processing/src/main/java/io/druid/segment/incremental/OnheapIncrementalIndex.java

```diff
@@ -333,4 +348,73 @@ public boolean compareCannonicalValues(String s1, String s2)
       return s1 == s2;
     }
   }
+
+  // Caches references to selector objetcs for each column instead of creating a new object each time in order to save heap space.
+  private static class ObjectCachingColumnSelectorFactory implements ColumnSelectorFactory
+  {
+    private final ConcurrentMap<String, LongColumnSelector> longColumnSelectorMap = Maps.newConcurrentMap();
```

Comment:
How about to use Cache in Guava, which also can specify various options, including expire policy.